### PR TITLE
Add Permission API

### PR DESCRIPTION
**Pull request by vyasswaroop**
- Added Permission APIs
    - `permission.create`
    - `permission.delete`
    - `permission.find`
    - `permission.current`
    - `permission.verify`
- Basic validation of requests
- Added Tests
- Updated to v1.2.0
- Updated Readme
- Added Placekey to Multifield parameters
- Updated documentation for placekey

[ch114003]

### CODE REVIEW

File: fc/fullcontact.go [hunk 1]
```diff
@@ -132,8 +132,12 @@ func sendToChannel(ch chan *APIResponse, response *http.Response, url string, er
...
 		case audienceCreateUrl, audienceDownloadUrl:
 			setAudienceResponse(apiResponse)
-		case emailVerificationUrl:
-			setEmailVerificationResponse(apiResponse)
+		case permissionFindUrl:
+			setPermissionFindResponse(apiResponse)
+		case permissionCurrentUrl:
+			setPermissionCurrentResponse(apiResponse)
+		case permissionVerifyUrl:
+			setPermissionVerifyResponse(apiResponse)
```
Comment (prakhar-j) [inline fc/fullcontact.go:140]: Where are you setting response for permission create and delete?

File: fc/fullcontact.go [hunk 2]
```diff
@@ -394,6 +398,127 @@ func (fcClient *fullContactClient) EmailVerification(email string) chan *APIResp
...
+		return ch
+	}
+	// Send Asynchronous Request in Goroutine
+	go fcClient.do(permissionCreateUrl, reqBytes, ch)
+	return ch
+}
+
+/* FullContact Permission API - PermissionDelete, takes an PermissionRequest and returns a channel of type APIResponse.
+Request is converted to JSON and sends a Asynchronous request */
+func (fcClient *fullContactClient) PermissionDelete(permissionRequest *MultifieldRequest) chan *APIResponse {
```
Comment (prakhar-j) [inline fc/fullcontact.go:428]: Can we call the variable multiFieldRequest. Same for find and current.

File: fc/fullcontact.go [hunk 3]
```diff
@@ -394,6 +398,127 @@ func (fcClient *fullContactClient) EmailVerification(email string) chan *APIResp
...
+	ch := make(chan *APIResponse)
+	if permissionRequest == nil {
+		go sendToChannel(ch, nil, "", NewFullContactError("Permission Request can't be nil"))
```
Comment (prakhar-j) [inline fc/fullcontact.go:431]: This message can use MultifieldRequest instead of Permission Request.

File: fc/consent.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
+package fullcontact
+
+type ConsentPurposesOption func(consentPurposes *ConsentPurposes)
+
+type ConsentPurposes struct {
```
Comment (prakhar-j) [inline fc/consent.go:5]: Can you include this code in permission_request.go, as now we have both consent.go and permission.go which can be confusing. As this is a request, we can include this in permission_request.go

File: fc/permission_request_test.go [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	assert.NoError(t, err)
+	requestJson := "{\"query\":{\"emails\":[\"marianrd97@outlook.com\",\"test1@gmail.com\",\"test2@outlook.com\"],\"phones\":[\"123-4567890\"],\"maids\":[\"abcd-123-abcd-1234-abcdlkjhasdfgh\",\"1234-snbk-lkldiemvmruixp-2kdp-vdm\"],\"location\":{\"addressLine1\":\"123/23\",\"addressLine2\":\"Some Street\",\"city\":\"Denver\",\"region\":\"Denver\",\"regionCode\":\"123123\",\"postalCode\":\"23124\"},\"name\":{\"given\":\"Marian\",\"family\":\"Reed\",\"full\":\"Marian C Reed\"},\"profiles\":[{\"url\":\"https://twitter.com/mcreedy\"},{\"url\":\"https://twitter.com/mcreedytest\"}]},\"consentPurposes\":[{\"purposeId\":1,\"channel\":[\"web\"],\"ttl\":365,\"enabled\":true}],\"locale\":\"US\",\"ipAddress\":\"127.0.0.1\",\"language\":\"en\",\"collectionMethod\":\"cookiePopUp\",\"collectionLocation\":\"Can we get a snapshot of where someone is opting in/out here?\",\"tcf\":\"some.valid.tcfv2.string\",\"policyUrl\":\"http://foo.baz\",\"termsService\":\"http://foo.tos\"}"
+	pr, err := NewPermissionRequest(
+		WithMultifieldRequestForPermission(multifieldRequest),
+		WithConsentPurposeForPermission(consentPurposes),
```
Comment (prakhar-j) [inline fc/permission_request_test.go:41]: Can you add a test with multiple consentPurposes, both valid and invalid

**Review by prakhar-j — CHANGES_REQUESTED** [review]
Thanks Vyas, Some changes required. Also need to add specific tests for placekey in person, resolve and permission. Also, need to add permission response tests.

File: README.md [hunk 6]
```diff
@@ -441,3 +450,160 @@ if resp.IsSuccessful {
...
+    - `Full`: _string_
+    - `Given`: _string_
+    - `Family`: _string_
+- `Profiles`: _[]*Profile_
+    - `Service`: _string_
+    - `Username`: _string_
+    - `Userid`: _string_
+    - `Url`: _string_
+- `Maids`: _[]string_
+- `Tags`: _[]*Tag_
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply README.md]: Removed

File: README.md [hunk 7]
```diff
@@ -441,3 +450,160 @@ if resp.IsSuccessful {
...
+- `query`: MultifieldReq - [required]
+- `consentPurposes`: List[ConsentPurposes] - [required]
+- `locale`: string
+- `ipAddress`: string
+- `language`: string
+- `collectionMethod`: string - [required]
+- `collectionLocation`: string - [required]
+- `policyUrl`: string - [required]
+- `termsService`: string - [required]
+-`tcf`: string
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply README.md]: Added space.

File: README.md [hunk 8]
```diff
@@ -441,3 +450,160 @@ if resp.IsSuccessful {
...
+- `PermissionFindResponse` - permission.find
+- `PermissionCurrentResponse` - permission.current
+
+`PermissionCreate` and `PermissionVerify` requires a `ResolveRequest` as parameter while the rest requires `MultifieldRequest` as parameter
+
+### Permission Create
+
+#### Parameters:
+Supported fields in query:
+- `query`: MultifieldReq - [required]
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply README.md]: Changed

File: example/main.go [hunk 9]
```diff
@@ -191,4 +191,71 @@ func main() {
...
+		fc.WithConsentPurposeId(1),
+		fc.WithConsentPurposeChannel("web"),
+		fc.WithConsentPurposeTtl(365),
+		fc.WithConsentPurposeEnabled(true))
+
+	permissionCreateRequest, err := fc.NewPermissionRequest(
+		fc.WithMultifieldRequestForPermission(multifieldRequest),
+		fc.WithConsentPurposeForPermission(consentPurposes),
+		fc.WithCollectionMethodForPermission("cookiePopUp"),
+		fc.WithCollectionLocationForPermission("Can we get a snapshot of where someone is opting in/out here?"),
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply example/main.go]: Changed

File: fc/consent.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+func validateConsentPurpose(consentPurpose *ConsentPurposes) error {
+	if consentPurpose.PurposeId == 0 {
+		return NewFullContactError("Purpose id is required for consentPurpose")
+	} else if consentPurpose.Channel == nil {
+		return NewFullContactError("Channel is required for consentPurpose")
+	} else if &consentPurpose.Enabled == nil {
+		return NewFullContactError("Enabled is required for consentPurpose")
+	} else if consentPurpose.Ttl == 0 {
+		return NewFullContactError("Ttl is required for consentPurpose")
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply fc/consent.go]: Removed check for ttl

File: fc/permission.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+	PermissionId   		string          	`json:"permissionId"`
+	ConsentPurposes    	[]*ConsentPurpose   `json:"consentPurposes"`
+	Locale     			string      		`json:"locale"`
+	IpAddress			string				`json:"ipAddress"`
+	Language     		string      		`json:"language"`
+	CollectionMethod	string				`json:"collectionMethod"`
+	CollectionLocation 	string      		`json:"collectionLocation"`
+	PolicyUrl     		string      		`json:"policyUrl"`
+	TermsService		string				`json:"termsService"`
+	Timestamp			string				`json:"timestamp"`
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply fc/permission.go]: changed timestamp to int

File: fc/fullcontact.go (hunk 1 above)
Comment (vyasswaroop) [reply fc/fullcontact.go:144]: Added response for create and delete

File: fc/fullcontact.go (hunk 2 above)
Comment (vyasswaroop) [reply fc/fullcontact.go]: changed

File: fc/fullcontact.go (hunk 3 above)
Comment (vyasswaroop) [reply fc/fullcontact.go:435]: changed

File: fc/consent.go (hunk 4 above)
Comment (vyasswaroop) [reply fc/consent.go]: Done. Moved the code to permission_request.go and also renamed it to ConsentPurpose

File: fc/permission.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	Created				int					`json:"created"`
+}
+
+type ConsentPurpose struct {
+	PurposeId		int			`json:"purposeId"`
+	Channel			string		`json:"channel"`
+	Ttl				int			`json:"ttl"`
+	Enabled			bool		`json:"enabled"`
+	AsOfTimestamp 	int			`json:"asOfTimestamp"`
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply fc/permission.go]: Updated the ConsentPurpose with purposeName and timestamp and using the same for PermissionCurrentResponse and PermissionVerifyResponse

File: fc/multifield_request.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+func (multifieldRequest *MultifieldRequest) isQueryable() bool {
+	return multifieldRequest.Emails != nil ||
+		multifieldRequest.Phones != nil ||
+		multifieldRequest.Profiles != nil ||
+		multifieldRequest.Maids != nil ||
+		isPopulated(multifieldRequest.LiNonId)
+}
+
+func (multifieldRequest *MultifieldRequest) validate() error {
```
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply fc/multifield_request.go:49]: Updated the validation

File: fc/permission_request_test.go (hunk 5 above)
Comment (vyasswaroop) [reply fc/permission_request_test.go:41]: Added test with consentPurpose

File: fc/fullcontact.go (hunk 3 above)
Comment (prakhar-j) [reply fc/fullcontact.go:435]: Error message is not updated

File: fc/fullcontact.go [hunk 14]
```diff
@@ -394,6 +402,127 @@ func (fcClient *fullContactClient) EmailVerification(email string) chan *APIResp
...
+	go fcClient.do(permissionDeleteUrl, reqBytes, ch)
+	return ch
+}
+
+/* FullContact Permission API - PermissionFind, takes an PermissionRequest and returns a channel of type APIResponse.
+Request is converted to JSON and sends a Asynchronous request */
+func (fcClient *fullContactClient) PermissionFind(multifieldRequest *MultifieldRequest) chan *APIResponse {
+	ch := make(chan *APIResponse)
+	if multifieldRequest == nil {
+		go sendToChannel(ch, nil, "", NewFullContactError("Permission Request can't be nil"))
```
Comment (prakhar-j) [inline fc/fullcontact.go:459]: Same here for error msg

File: fc/fullcontact.go [hunk 15]
```diff
@@ -394,6 +402,127 @@ func (fcClient *fullContactClient) EmailVerification(email string) chan *APIResp
...
+	go fcClient.do(permissionFindUrl, reqBytes, ch)
+	return ch
+}
+
+/* FullContact Permission API - PermissionCurrent, takes an PermissionRequest and returns a channel of type APIResponse.
+Request is converted to JSON and sends a Asynchronous request */
+func (fcClient *fullContactClient) PermissionCurrent(multifieldRequest *MultifieldRequest) chan *APIResponse {
+	ch := make(chan *APIResponse)
+	if multifieldRequest == nil {
+		go sendToChannel(ch, nil, "", NewFullContactError("Permission Request can't be nil"))
```
Comment (prakhar-j) [inline fc/fullcontact.go:483]: here too

File: fc/permission_request.go [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+	return nil
+}
+
+func validateForPermissionCreate(request *PermissionRequest) error {
+	err := validateForPermissionCreateFields(request)
+	if err != nil {
+		return err
+	}
+	for _, consentPurpose := range request.ConsentPurposes{
+		print(consentPurpose)
```
Comment (prakhar-j) [inline fc/permission_request.go:63]: There is a print statement. Can you search in the whole project for any other missed

File: fc/permission_request.go [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+	TermsService		string						`json:"termsService,omitempty"`
+	PurposeId			int							`json:"purposeId,omitempty"`
+	Channel				string						`json:"channel,omitempty"`
+}
+
+func NewPermissionRequest(option ...PermissionRequestOption) (*PermissionRequest, error) {
+	permissionRequest := &PermissionRequest{}
+
+	for _, opt := range option {
+		print(option)
```
Comment (prakhar-j) [inline fc/permission_request.go:27]: Print statement

File: fc/permission_request_test.go (hunk 5 above)
Comment (prakhar-j) [reply fc/permission_request_test.go:40]: I couldn't find it, can you point me to it.

File: fc/multifield_request.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (prakhar-j) [reply fc/multifield_request.go:49]: Need to add a case when only placekey is populated, and also its corresponding test

**Review by prakhar-j — CHANGES_REQUESTED** [review]
Some more changes required. Also tests for placekey need to be added for person and resolve, and I did not see any tests for permission response, Let me know if I am missing it somewhere.

File: fc/fullcontact.go (hunk 14 above)
Comment (vyasswaroop) [reply fc/fullcontact.go]: Updated the error message.

File: fc/fullcontact.go (hunk 15 above)
Comment (vyasswaroop) [reply fc/fullcontact.go]: Updated error message

File: fc/permission_request.go (hunk 16 above)
Comment (vyasswaroop) [reply fc/permission_request.go]: Removed print

File: fc/permission_request.go (hunk 17 above)
Comment (vyasswaroop) [reply fc/permission_request.go]: removed print

File: fc/multifield_request.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (vyasswaroop) [reply fc/multifield_request.go:48]: Added the check

File: fc/permission_request_test.go [hunk 18]
```diff
@@ -0,0 +1,642 @@
...
+		WithConsentPurposeForPermission(consentPurpose2),
+		WithCollectionMethodForPermission("cookiePopUp"),
+		WithCollectionLocationForPermission("Can we get a snapshot of where someone is opting in/out here?"),
+		WithPolicyUrlForPermission("http://foo.baz"),
+		WithTermsServiceForPermission("http://foo.tos"))
+	err := validateForPermissionCreate(pr)
+	assert.NoError(t, err)
+}
+
+func TestNewPermissionRequestForCreateWithMultipleConsentPurposeWithoutChannel(t *testing.T) {
```
Comment (prakhar-j) [inline fc/permission_request_test.go:369]: Test Name mentions "WithoutChannel" but channel is present.

File: fc/permission_request_test.go [hunk 19]
```diff
@@ -0,0 +1,642 @@
...
+	pr, _ := NewPermissionRequest(
+		WithMultifieldRequestForPermission(multifieldRequest),
+		WithConsentPurposeForPermission(consentPurpose1),
+		WithConsentPurposeForPermission(consentPurpose2),
+		WithCollectionMethodForPermission("cookiePopUp"),
+		WithCollectionLocationForPermission("Can we get a snapshot of where someone is opting in/out here?"),
+		WithPolicyUrlForPermission("http://foo.baz"),
+		WithTermsServiceForPermission("http://foo.tos"))
+	err := validateForPermissionCreate(pr)
+	assert.NoError(t, err)
```
Comment (prakhar-j) [inline fc/permission_request_test.go:395]: It should throw an error

**Review by prakhar-j — APPROVED** [review]
Just a minor fix required but Approving

File: fc/permission_request_test.go (hunk 18 above)
Comment (vyasswaroop) [reply fc/permission_request_test.go:369]: I'll change

File: fc/permission_request_test.go (hunk 19 above)
Comment (vyasswaroop) [reply fc/permission_request_test.go]: Updated

**Pull request merged**